spec/unit/govuk_index/popularity_worker_spec.rb

```diff
@@ -2,45 +2,73 @@
     ]
-    worker.perform(records, "govuk_test")
+    stub_document_lookups(documents)
+
+    document_ids = documents.map { |document| document["_id"] }
+    worker.perform(document_ids, "govuk_test")
 
-    expect(@processor).to have_received(:save).twice
+    expect(@processor).to have_received(:save).with(
+      having_attributes(
+        identifier: { "_id" => "document_1", "_version" => 1, "version_type" => "external_gte", "_type" => "generic-document" },
+        document: hash_including({ "popularity" => 0.7, "popularity_b" => 0.5, "title" => "test_doc1" }),
+      ),
+    ).once
+    expect(@processor).to have_received(:save).with(
+      having_attributes(
+        identifier: { "_id" => "document_2", "_version" => 1, "version_type" => "external_gte", "_type" => "generic-document" },
+        document: hash_including({ "popularity" => 0.6, "popularity_b" => 0.5, "title" => "test_doc2" }),
+      ),
+    ).once
     expect(@processor).to have_received(:commit)
   end
 
-  it "updates popularity field" do
-    stub_popularity_data("record_1" => { popularity_score: 0.7, popularity_rank: 0.5 })
-
-    @record = { "identifier" => { "_id" => "record_1" },
-                "document" => { "title" => "test_doc" } }
+  it "outputs to stdout and continues to the next document if a document is not found" do
+    stub_popularity_data({ "document_2" => { popularity_score: 0.6, popularity_rank: 0.5 } })
+    document = { "_id" => "document_2", "_source" => { "title" => "test_doc2" } }
+    allow(index).to receive(:get_document_by_id).with("document_1").and_return(nil)
+    allow(index).to receive(:get_document_by_id).with("document_2").and_return(document)
 
-    worker.perform([@record], "govuk_test")
+    document_ids = %w[document_1 document_2]
+    expect { worker.perform(document_ids, "govuk_test") }.to output("Skipping document_1 as it is not in the index\n").to_stdout
 
     expect(@processor).to have_received(:save).with(
       having_attributes(
-        identifier: { "_id" => "record_1", "version_type" => "external_gte", "_type" => "generic-document" },
-        document: hash_including({ "popularity" => 0.7, "popularity_b" => 0.5, "title" => "test_doc" }),
+        identifier: { "_id" => "document_2", "version_type" => "external_gte", "_type" => "generic-document" },
+        document: hash_including({ "popularity" => 0.6, "popularity_b" => 0.5, "title" => "test_doc2" }),
       ),
-    )
+    ).once
   end
 
   def stub_popularity_data(data = Hash.new({ popularity_score: 0.5, popularity_rank: 0.5 }))
     popularity_lookup = instance_double("popularity_lookup", lookup_popularities: data)
     allow(Indexer::PopularityLookup).to receive(:new).and_return(popularity_lookup)
   end
+
+  def stub_document_lookups(documents)
+    documents.each do |document|
+      allow(index).to receive(:get_document_by_id).with(document["_id"]).and_return(document)
+    end
```

Comment:
Minor nit as I'm not sure of practical implications.

Rather than setting up tons of individual stubs we could set up a single one:

```
allow(index).to receive(:get_document_by_id) { |document_id| documents[document_id] }
```